File: 21-shortest-path/solution.cpp
// Shortest Path in Unweighted Graph (BFS)
// Time: O(V + E), Space: O(V)

#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <unordered_set>

using namespace std;

class Solution {
public:
    int shortestPath(unordered_map<int, vector<int>>& graph, int start, int end) {
        if (start == end) return 0;
        
        queue<pair<int, int>> q;
        q.push({start, 0});
        unordered_set<int> visited;
        visited.insert(start);
        
        while (!q.empty()) {
            auto [node, dist] = q.front();
            q.pop();
            
            if (graph.find(node) != graph.end()) {
                for (int neighbor : graph[node]) {
                    if (neighbor == end) {
                        return dist + 1;
                    }
                    
                    if (visited.find(neighbor) == visited.end()) {
                        visited.insert(neighbor);
                        q.push({neighbor, dist + 1});
                    }
                }
            }
        }
        
        return -1;  // No path found
    }
};

// Test
int main() {
    Solution sol;
    unordered_map<int, vector<int>> graph;
    graph[0] = {1, 2};
    graph[1] = {0, 3, 4};
    graph[2] = {0, 4};
    graph[3] = {1};
    graph[4] = {1, 2, 5};
    graph[5] = {4};
    
    cout << sol.shortestPath(graph, 0, 5) << endl;  // 3
    cout << sol.shortestPath(graph, 0, 3) << endl;  // 2
    
    return 0;
}

File: 21-shortest-path/solution.java
// Shortest Path in Unweighted Graph (BFS)
// Time: O(V + E), Space: O(V)

import java.util.*;

public class Solution {
    public int shortestPath(Map<Integer, List<Integer>> graph, int start, int end) {
        if (start == end) return 0;
        
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{start, 0});
        Set<Integer> visited = new HashSet<>();
        visited.add(start);
        
        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            int node = curr[0];
            int dist = curr[1];
            
            for (int neighbor : graph.getOrDefault(node, new ArrayList<>())) {
                if (neighbor == end) {
                    return dist + 1;
                }
                
                if (!visited.contains(neighbor)) {
                    visited.add(neighbor);
                    queue.offer(new int[]{neighbor, dist + 1});
                }
            }
        }
        
        return -1;  // No path found
    }
    
    // Test
    public static void main(String[] args) {
        Solution sol = new Solution();
        Map<Integer, List<Integer>> graph = new HashMap<>();
        graph.put(0, Arrays.asList(1, 2));
        graph.put(1, Arrays.asList(0, 3, 4));
        graph.put(2, Arrays.asList(0, 4));
        graph.put(3, Arrays.asList(1));
        graph.put(4, Arrays.asList(1, 2, 5));
        graph.put(5, Arrays.asList(4));
        
        System.out.println(sol.shortestPath(graph, 0, 5));  // 3
        System.out.println(sol.shortestPath(graph, 0, 3));  // 2
    }
}
